---
title: "QTL analysis"
author: "Johan hallin"
date: "04 July 2016"
output: html_document
---
---

###Summary

This html contains the QTL analysis of the 10kcrosses project as well as heterosis and dominance calculations


Aneuploid Strains  
* 41
* 53
* 67
* 206
* 222
* 223
* 254
* 258 

-----

```{r global_options, include = FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
```

```{r dependencies}
library(qtl)
library(reshape2)
library(ggplot2)
library(stringr)
library(plyr)
library(pbapply)
```

```{r fixed variables}
##plots
sizeFacetTitles <- 12
sizeAxisLabels <- 12
sizeAxisTitle <- 12
axisAnglesX <- 45
sizeLegend <- 12
fontFamily <- 'Arial'

cb.palette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

##Calculations
alpha.value.qtl <- 0.10
heterosis.alpha.value <- 0.01
dominance.alpha.value <- 0.01
heterosis.alpha.values <- data.frame(p0.1 = 0.1, p0.075 = 0.075, p0.05 = 0.05, p0.025 = 0.025,
                                     p0.01 = 0.01, p0.006 = 0.006,p0.003 = 0.003, p0.0016 = 0.0016,
                                     p0.0008 = 0.0008, p0.0004 = 0.0004,
                                     p0.0002 = 0.0002)
dominance.alpha.values <- data.frame(p0.1 = 0.1, p0.075 = 0.075, p0.05 = 0.05, p0.025 = 0.025,
                                     p0.01 = 0.01, p0.006 = 0.006,p0.003 = 0.003, p0.0016 = 0.0016,
                                     p0.0008 = 0.0008, p0.0004 = 0.0004,
                                     p0.0002 = 0.0002)
```

-----

```{r function spreads}
#######
## Used to calculate the SE for all hybrids, used to create hybrid.SE
####
SE <- function(x) sd(x, na.rm = TRUE)/sqrt(length(x[!is.na(x)]))

CoV <- function(x) {
  mean <- mean(x, na.rm = TRUE)
  sd <- sd(x, na.rm = TRUE)
  cv <- sd / mean
  return(cv)
}

######
## Used in overDominancePhenoDistT.test
####
dist <- function(x, y) {
    sqrt(sum((x - y) ^ 2))
}
```

```{r function qtl maps paper}
######
## This function is used to show both yield and rate in the same plot
## Used to create QTL.maps
####
HomeMadeQtlMap <- function(environment, parent_or_residual) { #'parent' or 'residual'
  if(parent_or_residual == 'parent') {
    data <- parent_cross_OneDimScan
    thresholds <- parent.threshold
  } else {
    data <- cross_OneDimScan
    thresholds <- threshold
  }
  data2 <- data[grep(environment, names(data))]
  df <- data.frame(chr = cross_OneDimScan$chr, pos = cross_OneDimScan$pos,
                   gt = data2[grep('GT', names(data2))][[1]], yield = data2[grep('Yield', names(data2))][[1]])
  df$marker <- paste(df$chr, df$pos, sep = '-')
  df$marker <- factor(df$marker, levels = unique(df$marker))
  df.m <- melt(df, measure.vars =  c('gt', 'yield'))
  
  df.m.split <- split(df.m, df.m$chr)
  chr1 <- length(df.m.split$`1`[['chr']])/2 #/2 since there are 2 phenotypes
  chr2 <- chr1 + (length(df.m.split$`2`[['chr']])/2)
  chr3 <- chr2 + (length(df.m.split$`3`[['chr']])/2)
  chr4 <- chr3 + (length(df.m.split$`4`[['chr']])/2)
  chr5 <- chr4 + (length(df.m.split$`5`[['chr']])/2)
  chr6 <- chr5 + (length(df.m.split$`6`[['chr']])/2)
  chr7 <- chr6 + (length(df.m.split$`7`[['chr']])/2)
  chr8 <- chr7 + (length(df.m.split$`8`[['chr']])/2)
  chr9 <- chr8 + (length(df.m.split$`9`[['chr']])/2)
  chr10 <- chr9 + (length(df.m.split$`10`[['chr']])/2)
  chr11 <- chr10 + (length(df.m.split$`11`[['chr']])/2)
  chr12 <- chr11 + (length(df.m.split$`12`[['chr']])/2)
  chr13 <- chr12 + (length(df.m.split$`13`[['chr']])/2)
  chr14 <- chr13 + (length(df.m.split$`14`[['chr']])/2)
  chr15 <- chr14 + (length(df.m.split$`15`[['chr']])/2)
  chr16 <- chr15 + (length(df.m.split$`16`[['chr']])/2)
  
  rectangles <- data.frame(min = c(chr1, chr3, chr5, chr7,
                         chr9, chr11, chr13, chr15),
                  max = c(chr2, chr4, chr6, chr8,
                         chr10, chr12, chr14, chr16))
  
  gtThresholdNr <- grep(paste(environment, 'GT', sep = '_'), pheno.names)
  yieldThresholdNr <- grep(paste(environment, 'Yield', sep = '_'), pheno.names)
  df.m.gt <- df.m[df.m$variable == 'gt',]
  df.m.yield <- df.m[df.m$variable == 'yield',]
  
  genome.wide.fill <- 'grey90'
  genome.wide.color <- NA
  plot <- ggplot() + geom_point(data = df.m, size = 3,
                                aes(marker, value, shape = variable, group = variable)) +
    theme(axis.text.x = element_blank()) + scale_y_continuous(limits = c(-1, 25)) +
    geom_rect(data = rectangles, aes(xmin = min, xmax = max, ymin = 0, ymax = Inf),
              fill = genome.wide.fill, color = genome.wide.color, alpha = 0.5) +
    geom_point(data = df.m.gt[df.m.gt$value >= thresholds[gtThresholdNr],], aes(marker, value), color = 'red') +
    geom_point(data = df.m.yield[df.m.yield$value >= thresholds[yieldThresholdNr],], aes(marker, value),
               color = 'blue', shape = 17) +
    labs(x = NULL, y = 'LOD', title = paste(environment, parent_or_residual, sep = ' ')) +
    theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) + 
    theme(axis.text.y = element_text(size = sizeAxisLabels, family = fontFamily)) +
    theme(legend.title = element_blank()) +
    theme(legend.text = element_text(size = sizeLegend, family = fontFamily)) +
    theme(axis.ticks.x = element_blank())
  
  
  return(plot)
}

######
## This plots only one phenotype
####
HomeMadeQtlMapForPaper <- function(phenotype, parent_or_residual_lmm.dominance_or_LMM) { #'parent' or 'residual'
  if(parent_or_residual_lmm.dominance_or_LMM == 'parent') {
    data <- parent_cross_OneDimScan
    thresholds <- parent.threshold
  } else if(parent_or_residual_lmm.dominance_or_LMM == 'residual') {
    data <- cross_OneDimScan
    thresholds <- threshold
  } else if(parent_or_residual_lmm.dominance_or_LMM == 'lmm.dominance') {
    data <- lmm.dominance.df
    thresholds <- threshold
  } else if (parent_or_residual_lmm.dominance_or_LMM == 'LMM') {
    data <- lmm.df
    thresholds <- threshold
  }
  data2 <- data[grep(phenotype, names(data))]
  df <- data.frame(chr = data[['chr']], pos = data[['pos']], data2)
  df$marker <- paste(df$chr, df$pos, sep = '-')
  df$marker <- factor(df$marker, levels = unique(df$marker))
  names(df)[3] <- 'value'
  df.m <- df
  
  df.m.split <- split(df.m, df.m$chr)
  chr1 <- length(df.m.split$`1`[['chr']])
  chr2 <- chr1 + (length(df.m.split$`2`[['chr']]))
  chr3 <- chr2 + (length(df.m.split$`3`[['chr']]))
  chr4 <- chr3 + (length(df.m.split$`4`[['chr']]))
  chr5 <- chr4 + (length(df.m.split$`5`[['chr']]))
  chr6 <- chr5 + (length(df.m.split$`6`[['chr']]))
  chr7 <- chr6 + (length(df.m.split$`7`[['chr']]))
  chr8 <- chr7 + (length(df.m.split$`8`[['chr']]))
  chr9 <- chr8 + (length(df.m.split$`9`[['chr']]))
  chr10 <- chr9 + (length(df.m.split$`10`[['chr']]))
  chr11 <- chr10 + (length(df.m.split$`11`[['chr']]))
  chr12 <- chr11 + (length(df.m.split$`12`[['chr']]))
  chr13 <- chr12 + (length(df.m.split$`13`[['chr']]))
  chr14 <- chr13 + (length(df.m.split$`14`[['chr']]))
  chr15 <- chr14 + (length(df.m.split$`15`[['chr']]))
  chr16 <- chr15 + (length(df.m.split$`16`[['chr']]))
  
  rectangles <- data.frame(min = c(chr1, chr3, chr5, chr7,
                         chr9, chr11, chr13, chr15),
                  max = c(chr2, chr4, chr6, chr8,
                         chr10, chr12, chr14, chr16))
  
  gtThresholdNr <- grep(phenotype, pheno.names)
  
  genome.wide.fill <- 'grey90'
  genome.wide.color <- NA
  plot <- ggplot() + geom_line(data = df.m,
                                aes(marker, value, group = 1)) +
    theme_classic() +
    geom_point(data = df.m[df.m$value >= thresholds[gtThresholdNr],], 
              aes(marker, value), color = 'red') +
    theme(axis.text.x = element_blank()) + scale_y_continuous(limits = c(0, 25)) +
    geom_rect(data = rectangles, aes(xmin = min, xmax = max, ymin = 0, ymax = 25),
              fill = genome.wide.fill, color = genome.wide.color, alpha = 0.5) +
    labs(x = NULL, y = 'LOD', title = paste(phenotype, parent_or_residual_lmm.dominance_or_LMM, sep = ' ')) +
    theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) + 
    theme(axis.text.y = element_text(size = sizeAxisLabels, family = fontFamily)) +
    theme(legend.title = element_blank()) +
    theme(legend.text = element_text(size = sizeLegend, family = fontFamily)) +
    theme(axis.ticks.x = element_blank())
  
  
  return(plot)
}
```

```{r function plot phenotype marker distribution}
######
## This function creates plots to look at the distribution of genotypes at a given marker
## It also provides numbers for genotypeHeterosis by setting output to geno.pheno. See 
## the function genotypeHeterosis for more info.
####
PhenotypeMarkerDistribution <- function(phenotype, marker, output) {
  pheno.value <- original.median[original.median$phenotype == phenotype,] #These are the original phenotype values, not the residuals,
  pheno.value <- data.frame(id = pheno.value$strain, pheno.value = pheno.value$value)
  geno.value <- geno[,grep(paste('id', marker, sep = '|'), names(geno))]
  geno.pheno <- merge(geno.value, pheno.value, by = 'id')
  geno.pheno.m <- na.exclude(melt(geno.pheno, measure.vars = 
                                   gsub('?_.*$', '', marker)))
  geno.pheno.s <- split(geno.pheno.m, geno.pheno.m$variable)
  geno.pheno.m <- do.call(rbind, geno.pheno.s)

  marker.pheno.plot <- ggplot(geno.pheno.m, aes(value, pheno.value, 
                                             fill = value)) +
    scale_fill_manual(values=cb.palette) + geom_boxplot() + theme_bw() +
    facet_wrap(~variable, ncol = 4) +
    labs(y = NULL, x = NULL, title = paste(phenotype, marker, sep =' ')) +
    theme(legend.position = 'none') +
    theme(axis.title.x = element_text(size = sizeAxisTitle, family = fontFamily)) +
    theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) + 
    theme(axis.text.x = element_text(size = sizeAxisLabels, family = fontFamily)) +
    theme(axis.text.y = element_text(size = sizeAxisLabels, family = fontFamily)) +
    theme(strip.text.x = element_text(size = sizeFacetTitles, family = fontFamily))
  
  geno.pheno.m$pheno <- phenotype
  
  if(output == 'plot') {
    return(marker.pheno.plot)
  } else if(output == 'geno.pheno') {
    return(geno.pheno)
  }
}
```

```{r function phased genotypes genotype phenotype}
######
## This plot connects the phenotypes with the phased genotypes
####
plotPhasedGenotypes <- function(x) {
  phenotype <- x[['phenotype']]
  marker <- x[['marker']]
  
  geno <- phased.hybrid.genotypes[phased.hybrid.genotypes$id == marker,]
  geno.t <- melt(geno, id.vars = 'id')
  names(geno.t) <- c('marker', 'strain', 'genotype')
  pheno.values <- original.median[original.median$phenotype == phenotype,]
  pheno.values$id <- as.character(pheno.values$strain)
  data <- join(pheno.values, geno.t, by = 'strain', type = 'inner')
  
  ttest <- t.test(x = data[data$genotype == 'NW',]$value, y = data[data$genotype == 'WN',]$value, paired = FALSE, var.equal = FALSE)
  
  plot <- ggplot(data, aes(genotype, value, fill = genotype)) + 
    scale_fill_manual(values = cb.palette[2:8]) + geom_violin()
  
  out <- list(plot = plot, ttest = ttest)
  
  return(out)
}
```

```{r function phenotype distribution comparisons}
######
## This function creates the data needed to make plots to compare
## the F12 parents, the POLs and the estimated parents
## Using original.median and F12seg.crossdata for this
####
phenoDistCompF12segHybrid <- function(environment, pheno) {
  if(pheno == 'generationTime') {
    F12pheno <- -pull.pheno(F12seg.crossdata, pheno.col = grep(environment, names(F12seg.crossdata$pheno)))
    F12pheno <- data.frame(value = F12pheno, ura = F12seg.crossdata$pheno$ura, id = F12seg.crossdata$pheno$id,
                           dataset = 'F12 segregants', environment = environment)
    F12pheno[F12pheno$ura == 1,]$ura <- 'ura+'
    F12pheno[F12pheno$ura == 0,]$ura <- 'ura-'
    est.parents <- parent_pheno[,grep('GT', names(parent_pheno))]
    est.parents <- est.parents[,grep(environment, names(est.parents))]
    est.parents <- data.frame(value = est.parents, ura = F12pheno$ura, id = parent_pheno$id,
                              dataset = 'Estimated Parents', environment = environment)
    Hybrid <- original.median[original.median$variable != 'Yield',]
    Hybrid <- Hybrid[grep(environment, Hybrid$environment),]
    Hybrid <- data.frame(value = Hybrid$value, ura = 'Hybrids', id = Hybrid$strain, dataset = 'Hybrids', 
                        environment = environment)
  
    data <- rbind(F12pheno, Hybrid, est.parents)
  
    return(data) 
  } else if(pheno == 'yield') {
    F12pheno <- -pull.pheno(F12seg.crossdata, pheno.col = grep(environment, names(F12seg.crossdata$pheno)))
    F12pheno <- data.frame(value = F12pheno, ura = F12seg.crossdata$pheno$ura, id = F12seg.crossdata$pheno$id,
                           dataset = 'F12 segregants', environment = environment)
    F12pheno[F12pheno$ura == 1,]$ura <- 'ura+'
    F12pheno[F12pheno$ura == 0,]$ura <- 'ura-'
    
    est.parents <- parent_pheno[,grep('Yield', names(parent_pheno))]
    est.parents <- est.parents[,grep(environment, names(est.parents))]
    est.parents <- data.frame(value = est.parents, ura = F12pheno$ura, id = parent_pheno$id,
                              dataset = 'Estimated Parents', environment = environment)
    Hybrid <- original.median[original.median$variable == 'Yield',]
    Hybrid <- Hybrid[grep(environment, Hybrid$environment),]
    Hybrid <- data.frame(value = Hybrid$value, ura = 'Hybrids', id = Hybrid$strain, dataset = 'Hybrids', 
                        environment = environment)
  
    data <- rbind(Hybrid, est.parents)
  
    return(data)
  }
}

```

```{r function extract peaks}
######
## Extract QTLs 
####
extractPeaks <- function(phenotype, dataSet) { #dataset = parent or residual
  lodcol <- grep(phenotype, pheno.names)
  if(dataSet == 'residual') {
    threshold <- threshold[lodcol]
    peaks <- summary(cross_OneDimScan, lodcolumn = lodcol, threshold = threshold)
  } else if(dataSet == 'parent') {
    threshold <- parent.threshold[lodcol]
    peaks <- summary(parent_cross_OneDimScan, lodcolumn = lodcol, threshold = threshold)
  }
  if(length(peaks[[phenotype]]) >= 1) {
    df <- data.frame(phenotype = phenotype, marker = row.names(peaks), chr = peaks$chr,
                     pos = peaks$pos, LOD = peaks[[phenotype]])
    df$marker <- gsub('\\.', '-', df$marker)
  }
  df <- df[df$marker != 'NA',]
  return(df)
}

######
## Finalize QTL extraction with lod interval and layout
####
peakFinalize <- function(peaks, n) {
  peaks <- do.call(rbind, peaks)
  peaks2 <- apply(peaks, 1, lodInterval, n = n)
  peaks2 <- data.frame(do.call(rbind, peaks2))
  peaks2 <- peaks2[names(peaks2) != 'chr' & names(peaks2) != 'pos']
  peaks2$intLength <- as.numeric(peaks2$intLength)
  intLenMedian <- median(peaks2$intLength)
  peaks2$LOD <- as.numeric(peaks2$LOD)
  
  peaks2$VarianceExplained <- 1 - 10 ** (-2 * (peaks2$LOD/n))
  peaks2 <- peaks2[,-c(4,5)]
  if(n > 5000) {
    peaks2$dataset <- 'residual'
  } else {
    peaks2$dataset <- 'parental'
  }
  
  return(peaks2)
}

######
## Used in peakFinalize
## Uses the 'peaks' df in order to pick out the QTLs and the calculate the lod interval
####
lodInterval <- function(x, n) {
  pheno <- x[['phenotype']]
  if(n < 5000) {
    lodcol <- grep(x[['phenotype']], names(parent_cross_OneDimScan)) - 2 #-2 since the df starts with chr and pos
    intdf <- try(lodint(parent_cross_OneDimScan, chr = x[['chr']], lodcolumn = lodcol, drop = 1.8), silent = T)
  } else {
    lodcol <- grep(x[['phenotype']], names(cross_OneDimScan)) - 2 #-2 since the df starts with chr and pos
    intdf <- try(lodint(cross_OneDimScan, chr = x[['chr']], lodcolumn = lodcol, drop = 1.8), silent = T)
  }
  
  if(length(intdf) > 1) {
      int <- intdf[['pos']][c(1,3)] #Leaving the peak position out
      x$intStart <- int[1]
      x$intEnd <- int[2]
      x$intLength <- int[2] - int[1]
  } else {
      x$intStart <- NA
      x$intEnd <- NA
      x$intLength <- NA
  }
  return(x)
}

######
## Adds the QTLs I picked out myself
####
addQTLS <- function(x) {
  y <- x
  if(names(y) == 'allantoin_GTT') names(y) <- 'allantoin_GT'
  lod <- cross_OneDimScan[row.names(cross_OneDimScan) == paste('chr', y[[1]], sep =''),]
  LOD <- lod[,names(lod) == names(y)]
  chr <- gsub('?\\..*$', '', y[[1]])
  pos <- gsub('^.*?\\.', '', y[[1]])
  marker <- paste('chr', y[[1]], sep = '')
  marker <- gsub('\\.', '-', y[[1]])
  list <- list(data.frame(phenotype = names(y), marker, chr, pos, LOD))
  names(list) <- names(y)
  return(list)
}
```

```{r function additional QTLs}
######
## This function aided me in finding additional QTLs. It doesn't pick out marker at the peak, it gives 
## a number of markers, have to get the highest one yourself. R/qtl only takes highest peak on a
## given chr. It's used to create multiple.qtls
####

MultiQTL <- function(phenotype) {
  thresholdNr <- grep(phenotype, pheno.names)
  df <- data.frame(chr = cross_OneDimScan$chr, pos = cross_OneDimScan$pos,
                   value = cross_OneDimScan[[phenotype]])
  df <- df[MultiQTLNarrowDown(df$value),]
  df <- df[df$value >= threshold[thresholdNr],]
  
  return(df)
}

##Used in MultipleQTL above
MultiQTLNarrowDown <- function(x) {
  # Use -Inf instead if x is numeric (non-integer)
  y <- diff(c(-.Machine$integer.max, x)) > 0L
  rle(y)$lengths
  y <- cumsum(rle(y)$lengths)
  y <- y[seq.int(1L, length(y), 2L)]
  if (x[[1]] == x[[2]]) {
    y <- y[-1]
  }
  return(y)
}
```

```{r function to calculate heterosis}
heterosisAnalysis <- function(alpha.value, centralTendency) {
  print(alpha.value)
  
  hybrids <- hybrid.raw.clean.2way
  hybrids$a <- gsub('?-.*$', '', hybrids$strain)
  hybrids$alpha <- gsub('^.*?-', '', hybrids$strain)
  
  hybrids.pheno.s <- split(hybrids, hybrids$phenotype)
  
  heterosisanalysis.env <- new.env()
  heterosisanalysis.env$centralTendency <- centralTendency
  heterosisanalysis.env$alpha.value <- alpha.value
  
  calculations <- function(x, env) {
    print(unique(x$phenotype))
    x.s <- split(x = x, f = x[['strain']])

    ##Remove hybrids with one or zero phenotype values
    rm.list <- lapply(x.s, FUN = function(x) length(na.exclude(x[['value']])) <= 1)
    rm.list <- do.call(rbind, rm.list)
    hybrids.s <- x.s[!as.vector(rm.list)]
    
    calculation.env <- new.env()
    calculation.env$hybrids <- x
    calculation.env$alpha.value <- env$alpha.value
    calculation.env$centralTendency <- env$centralTendency
    result <- pblapply(hybrids.s, heterosisCalculation, env = calculation.env)
    
    return(result)
  }
  
  results <- pblapply(hybrids.pheno.s, calculations, env = heterosisanalysis.env)
  
  return(results)
}
  
heterosisCalculation <- function(x, env) {

  parent.a <- env$hybrids[env$hybrids[['a']] == unique(x[['a']]),]
  parent.alpha <- env$hybrids[env$hybrids[['alpha']] == unique(x[['alpha']]),]
  parent.a$value[parent.a$value == -Inf] <- NA
  parent.alpha$value[parent.alpha$value == -Inf] <- NA
  
  heterosiscal.env <- new.env()
  heterosiscal.env$alpha.value <- env$alpha.value
  heterosiscal.env$centralTendency <- env$centralTendency
  heterosiscal.env$parent.a <- parent.a
  heterosiscal.env$parent.alpha <- parent.alpha
  
  test.result <- heterosisCall(x = x, env = heterosiscal.env)
  
  result <- heterosisAssign(x = test.result)
  
  return(result)
}

heterosisCall <- function(x, env) {
  hybrid.average <- env$centralTendency(x[['value']], na.rm = TRUE)
  parent.average <- data.frame(parent = c('parent.a', 'parent.alpha'),
                               value = c(env$centralTendency(env$parent.a$value, na.rm = TRUE),
                               env$centralTendency(env$parent.alpha$value, na.rm = TRUE)))
  stopifnot(parent.average$value[1] != parent.average$value[2])
  
  strong.parent <- as.character(parent.average[parent.average$value == max(parent.average$value),]$parent)
  weak.parent <- as.character(parent.average[parent.average$value == min(parent.average$value),]$parent)
  
  
  heteroiscall.env <- new.env()
  heteroiscall.env$parent.a <- env$parent.a
  heteroiscall.env$parent.alpha <- env$parent.alpha
  heteroiscall.env$parent.average <- parent.average
  heteroiscall.env$hybrid <- x
  heteroiscall.env$hybrid.average <- hybrid.average
  heteroiscall.env$parent.average <- parent.average
  heteroiscall.env$alpha.value <- env$alpha.value
  heteroiscall.env$strong.parent.name <- strong.parent
  heteroiscall.env$weak.parent.name <- weak.parent
  heteroiscall.env$strong.parent <- env[[strong.parent]]
  heteroiscall.env$weak.parent <- env[[weak.parent]]
  
  mid.parent <- modelMP(round = 2, env = heteroiscall.env)
  mid.parent.average <- env$centralTendency(mid.parent, na.rm = TRUE)
  
  heteroiscall.env$mid.parent <- mid.parent
  heteroiscall.env$mid.parent.average <- mid.parent.average
  
  BPH <- testForBPH(env = heteroiscall.env)
  WPH <- testForWPH(env = heteroiscall.env)
  AsMP <- testForAsMP(env = heteroiscall.env)
  PosMPH <- testForPosMPH(env = heteroiscall.env)
  NegMPH <- testForNegMPH(env = heteroiscall.env)
  NonSigParDif <- testForNotSignificantParentDifferences(env = heteroiscall.env)
  
  test.result <- c(BPH, WPH, AsMP, PosMPH, NegMPH, NonSigParDif)
      
  return(test.result)
}
    
    
heterosisAssign <- function(x) {
  test.result.clean <- x[which(x != 'ambiguous')]
  
  if(length(which(test.result.clean == 'notSignificantParentDifference')) == 1) {
    heterosis <- 'notSignificantParentDifference'
  } else if(length(which(test.result.clean == 'notSignificantParentDifference')) == 0) {
    if(length(test.result.clean) >= 3) {
      heterosis <- 'ambiguous'
    } else if(length(test.result.clean) <= 2 & length(test.result.clean) >= 1) {
      heterosis <- paste(test.result.clean, collapse = '|')
    } else { heterosis <- 'ambiguous' }
  }
  
  return(heterosis)
}
    
testForBPH <- function(env) {
  hybrid.values <- env$hybrid[['value']]
  hybrid.values <- hybrid.values[hybrid.values != -Inf]
  ttest <- tTest(x = env[['strong.parent']][['value']], y = hybrid.values)
  p.value <- ttest[[3]]
  
  if(p.value <= env$alpha.value) {
    if(env$hybrid.average > env$parent.average[env$parent.average$parent == env$strong.parent.name,][['value']]) {
      heterosis <- 'BPH'
    } else { heterosis <- 'ambiguous' }
  } else { heterosis <- 'ambiguous' }
  
  return(heterosis)
}

testForWPH <- function(env) {
  hybrid.values <- env$hybrid[['value']]
  hybrid.values <- hybrid.values[hybrid.values != -Inf]
  ttest <- tTest(x = env[['weak.parent']][['value']], y = hybrid.values)
  p.value <- ttest[[3]]
  
  if(p.value <= env$alpha.value) {
    if(env$hybrid.average < env$parent.average[env$parent.average$parent == env$weak.parent.name,][['value']]) {
      heterosis <- 'WPH'
    } else { heterosis <- 'ambiguous' }
  } else { heterosis <- 'ambiguous' }
  
  return(heterosis)
}

testForPosMPH <- function(env) {
  hybrid.values <- env$hybrid[['value']]
  hybrid.values <- hybrid.values[hybrid.values != -Inf]
  ttest <- tTest(x = env[['mid.parent']], y = hybrid.values)
  p.value <- ttest[[3]]
  
  if(p.value <= env$alpha.value) {
    if(env[['hybrid.average']] > env[['mid.parent.average']]) {
      heterosis <- 'PosMPH'
    } else { heterosis <- 'ambiguous' }
  } else { heterosis <- 'ambiguous' }
  
  return(heterosis)
}

testForNegMPH <- function(env) {
  hybrid.values <- env$hybrid[['value']]
  hybrid.values <- hybrid.values[hybrid.values != -Inf]
  ttest <- tTest(x = env[['mid.parent']], y = hybrid.values)
  p.value <- ttest[[3]]
  
  if(p.value <= env$alpha.value) {
    if(env[['hybrid.average']] < env[['mid.parent.average']]) {
      heterosis <- 'NegMPH'
    } else { heterosis <- 'ambiguous' }
  } else { heterosis <- 'ambiguous' }
  
  return(heterosis)
}

testForAsMP <- function(env) {
  hybrid.values <- env$hybrid[['value']]
  hybrid.values <- hybrid.values[hybrid.values != -Inf]
  ttest.strong.parent <- tTest(x = env[['strong.parent']][['value']], y = hybrid.values)
  ttest.weak.parent <- tTest(x = env[['weak.parent']][['value']], y = hybrid.values)
  ttest.mid.parent <- tTest(x = env[['mid.parent']], y = hybrid.values)
  
  parent.p.values <- c(ttest.strong.parent[[3]], ttest.weak.parent[[3]])
  parent.significant <- parent.p.values[which(parent.p.values < env$alpha.value)]
  
  if(length(parent.significant) == 2) {
    if(ttest.mid.parent[[3]] > env$alpha.value) {
      heterosis <- 'AsMP'
    } else { heterosis <- 'ambiguous' }
  } else { heterosis <- 'ambiguous' }
  
  return(heterosis)
}

testForNotSignificantParentDifferences <- function(env) {
  ttest.strong.vs.weak.parent <- tTest(x = env[['strong.parent']][['value']], y = env[['weak.parent']][['value']])
  
  if(ttest.strong.vs.weak.parent[[3]] > env$alpha.value) {
    heterosis <- 'notSignificantParentDifference'
  } else { heterosis <- 'ambiguous' }
  
  return(heterosis)
}

modelMP <- function(round, env) {
  parent.average <- mean(env$parent.average$value)
  repeat{
    mp.modelled <- rnorm(n = round(mean(length(env$parent.a$value), length(env$parent.alpha$value)), 0),
                          mean = parent.average, sd = max(sd(x = env$parent.a$value, na.rm = TRUE), sd(env$parent.alpha$value, na.rm = TRUE)))
    mp.modelled.mean <- mean(mp.modelled)
    mp.modelled.sd <- sd(mp.modelled)
    if(round(mp.modelled.mean, digits = round) == round(parent.average, digits = round) &
       round(mp.modelled.sd, digits = round) == round(max(sd(x = env$parent.a$value, na.rm = TRUE), sd(env$parent.alpha$value, na.rm = TRUE)), digits = round)) break
  }
  return(mp.modelled)
}

tTest <- function(x, y) {
  out <- t.test(x = x, y = y, var.equal = FALSE, paired = FALSE)
  return(out)
}
```

```{r function heterosis alpha threshold summary}
######
## This function makes a summary of the comparison of different alpha thresholds
####
heterosisDataAlphaComparisonSummary <- function(x) {
  nonSignificant <- length(x[grep('notSignificantParentDifference|ambiguous', x[['value']]),][['value']]) / length(x[['value']])
  dataSinNonSig <- x[-grep('notSignificantParentDifference|ambiguous', x[['value']]),]
  tot.ln <- length(dataSinNonSig$value)
  BPH <- length(dataSinNonSig$value[dataSinNonSig$value == 'BPH|PosMPH']) / tot.ln
  WPH <- length(dataSinNonSig$value[dataSinNonSig$value == 'WPH|NegMPH']) / tot.ln
  AsMP <- length(dataSinNonSig$value[dataSinNonSig$value == 'AsMP']) / tot.ln
  NegMPH <- length(dataSinNonSig$value[dataSinNonSig$value == 'NegMPH']) / tot.ln
  NegMPH <- NegMPH + WPH
  PosMPH <- length(dataSinNonSig$value[dataSinNonSig$value == 'PosMPH']) / tot.ln
  PosMPH <- PosMPH + BPH
  df <- data.frame(pValue = unique(x[['L1']]), nonSignificant, BPH, WPH, AsMP, PosMPH, NegMPH)
}
```

```{r function heterosisVsDominance}
######
## This function creates the heterosis.vs.dominance list of things which means 
## it creates the plot comparing BPH, All POLs and WPH used in the paper, as well
## as the data for that plot
####
HeterosisVsDominance <- function(phenotype) {
  names(heterosis.data) <- c('heterosis', 'id', 'phenotype', 'alpha.value')
  het <- heterosis.data[heterosis.data$phenotype == phenotype,]
                       
  het <- het[grep('BPH|WPH', het$heterosis),]
  het <- het[,grep('heterosis|id', names(het))]
  het$heterosis <- gsub('?\\|.*$', '', het$heterosis)
  
  #Not using the parental markers that are also in the residuals, still split at phenotypes though.
  #Taken the parental QTLs from the QTL_summaries.xlsx file
  markers <- total.peaks[total.peaks$dataset != 'parental',]
  markers2 <- rbind(markers,
                    total.peaks[grep('chr3-240075|chr7-487435|chr13-389211|chr15-1006391|chr15-991945|chr16-680182',
                                    total.peaks$marker),])
  
   results <- sigMark(phenotypeName = phenotype, output = 'highlight',
                     highlight_df = het, marker = markers2)
  
  return(results)
}
```

```{r function dominance chi square}
######
## This function does the chi square test to check for overrepresentation of 
## heterozygotes or homozyogotes within the BPH or WPH compared to the entire population
## Don't see if they are enriched or depleted, only that they are different
####
##x = the comparisonDFchi data in heterosis.vs.dominance
chiSquareTest <- function(x) {
  y <- split(x, list(x$heterosis, x$dataset))
  p.values <- lapply(y, chiSq)
  out <- do.call(rbind, p.values)
  return(out)
}

##Function for implementing chiSquare, used in chiSquareTest function
chiSq <- function(p) {
  DF <- 1
  WW <- chiSquare(observed = p$heterosisStrains[1], expected = p$expected[1])
  WWrest <- chiSquare(observed = sum(p$heterosisStrains[-1]), expected = sum(p$expected[-1]))
  chi2WW <- sum(WW, WWrest)
  pWW <- pchisq(chi2WW, DF, lower.tail=FALSE) 
  
  NW <- chiSquare(observed = p$heterosisStrains[2], expected = p$expected[2])
  NWrest <- chiSquare(observed = sum(p$heterosisStrains[-2]), expected = sum(p$expected[-2]))
  chi2NW <- sum(NW, NWrest)
  pNW <- pchisq(chi2NW, DF, lower.tail=FALSE) 
  
  NN <- chiSquare(observed = p$heterosisStrains[3], expected = p$expected[3])
  NNrest <- chiSquare(observed = sum(p$heterosisStrains[-3]), expected =  sum(p$expected[-3]))
  chi2NN <- sum(NN, NNrest)
  pNN <- pchisq(chi2NN, DF, lower.tail=FALSE)
  
  pValues <- data.frame(WW = pWW, NW = pNW, NN = pNN, marker = p$marker[1], type = p$heterosis[1],
                        dataset = p$dataset[1])
  return(pValues)
}

##Function for chi square test, used in chiSq
chiSquare <- function(observed, expected) {
  deviation <- observed - expected
  deviation.sq <- deviation ** 2
  chi.sq <- deviation.sq / expected
  return(chi.sq)
}
```

```{r function dominance calling}
######
## I'll be using chiSq.df for this, that is where I have the chi square information
## for all QTLs and all genotypes at the QTL. The definition is the following:
  # Overdominance:
    # Enriched heterozygotes while depleted or unchanged homozygotes
  # Dominance:
    # Enriched strong homozygote while overrepresentation or unchanged heterozygote (for BPH, opposite for WPH)
  # homozygous.genotype.enriched
    # This is done in the calculation for Dominance, and it is simply when the heterozygote is depleted but
    # the strong homozygote is still enriched, not technically dominance, but still interesting which is 
    # why it gets it's own designation. To get all accounts where the homozygote is enriched this needs to be
    # (regardless of the status of the heterozygous) added to the values for Dominance
## I'll apply this function to the rows of chiSq.df, so x = row of chiSq.df
####

######
## you get a data.frame out of this, I will explain the different columns
## change = the frequency of the genotypes in the subset of the POLs (WPH or BPH)
## original = the frequency of the genotypes in all POLs
## chi.square = the p-value for the chi.square test to see if the frequencies between change and original differ
## status = whether the genotype is enriched or depleted, if no difference between frequencies it = unchanged
## heterosis = the subset of the population considered. BPH or WPH
## marker = the QTL considered. The QTLs used are the unique to one phenotype, i.e. I don't take both parental
 # and residual QTLs that come out in the same phenotype
## phenotype = the phenotype considered
## dataset = where the QTL was called, residual or parental. Some of the residual QTLs used may also have been called
 # in the parental scan, but I only take them once
## pheno.value = this is the average phenotype for all the POLs depending on the genotype
## pheno.strength = this is the relative strength of the genotype's phenotype. It is either the weak, strong or middle
 # phenotype. Middle is divided into middle.weak, for a middle phenotype closer to the weak one, and middle.strong
 # for a middle phenotype closer to the strong. If precisly in the middle it = middle. There is now statistical test
 # done here, so whether it is significantly closer to strong than to weak, I do not take into account, and it doesn't
 # matter at this point.
####

##Master function
dominanceCalling <- function(x, alpha) { #x = row of chiSq.df
  alpha.value <- alpha
  compare.fractions <- heterosis.vs.dominance[[as.character(x[['pheno']])]]$comparisonDf[[as.character(x[['marker']])]]
  x.m <- melt(x)
  enrich.deplete.df <- dominanceConnectData(type = as.character(x[['type']]))
  enrich.phenotypes.strength <- dominanceConnectPhenotype(x = enrich.deplete.df)
  enrich.phenotypes.dominance <- dominanceAssign(x = enrich.phenotypes.strength, type = as.character(x[['type']]))
  return(enrich.phenotypes.dominance)
}

##Used in dominanceCalling
dominanceConnectData <- function(type) {
  alpha.value <- get('alpha.value', parent.frame())
  compare.fractions <- get('compare.fractions', parent.frame())
  x.m <- get('x.m', parent.frame())
  
  if(type == 'WPH') type <- 'WPH.WPH' #Due to the name in compare.fractions
  df <- data.frame(change = compare.fractions[[type]], original = compare.fractions[['original']], 
                   genotype = compare.fractions[['genotype']], chi.square = x.m[['value']])
  
  change <- apply(df, 1, dominanceChangeInFrequency, alpha = alpha.value)
  df[['status']] <- change
  df[['heterosis']] <- gsub('?\\..*$', '', type) #Get rid of .WPH when type = WPH
  df[['marker']] <- x.m[['marker']]
  df[['phenotype']] <- x.m[['pheno']]
  df[['dataset']] <- x.m[['dataset']]
  return(df)
}

##Used in dominanceConnectData
dominanceChangeInFrequency <- function(y, alpha) { #y = row of df
  change <- NULL
  if(as.numeric(y[['chi.square']]) > alpha) { #Have to put as.numeric since apply makes it into character
    change <- 'unchanged'
  } else {
    if(as.numeric(y[['change']]) < as.numeric(y[['original']])) {
      change <- 'depleted'
    } else {
      change <- 'enriched'
    }
  }
  return(change)
}

##Used in dominanceCalling
dominanceConnectPhenotype <- function(x) { #Connecting to actual phenotype
  pheno.for.geno <- sigMark(phenotypeName = unique(x[['phenotype']]),
                            output = 'df', marker = NA, highlight_df = NULL)
  pheno.for.geno <- pheno.for.geno[pheno.for.geno[['marker']] == unique(x[['marker']]) &
                                   pheno.for.geno[['dataset']] == unique(x[['dataset']]),]
  
  stopifnot(length(pheno.for.geno$genotype) == 3)
  
  enrich.phenotypes.df <- join(x = x,
                               y = pheno.for.geno[c('genotype', 'pheno.value')], by = 'genotype')
  enrich.phenotypes.df$pheno.strength <- NA #Adding empty column to put the strength of the phenotype
  
  ##Assigning the genotypes depending on how strong they are compared to the others
  enrich.phenotypes.df[enrich.phenotypes.df$pheno.value == min(enrich.phenotypes.df$pheno.value),]['pheno.strength'] <- 'weak'
  enrich.phenotypes.df[enrich.phenotypes.df$pheno.value == max(enrich.phenotypes.df$pheno.value),]['pheno.strength'] <- 'strong'
  enrich.phenotypes.df[enrich.phenotypes.df$pheno.value == median(enrich.phenotypes.df$pheno.value),]['pheno.strength'] <- 'middle'
  
  ##Assign the middle as closer to the weak or strong genotype
  additive.value <- mean(c(enrich.phenotypes.df[enrich.phenotypes.df[['pheno.strength']] == 'strong',][['pheno.value']], 
                          enrich.phenotypes.df[enrich.phenotypes.df[['pheno.strength']] == 'weak',][['pheno.value']]))
  
  middle.value <- enrich.phenotypes.df[enrich.phenotypes.df$pheno.strength == 'middle',]$pheno.value
  
  if(middle.value == additive.value) {
  } else if(middle.value < additive.value) {
    enrich.phenotypes.df[enrich.phenotypes.df$pheno.strength == 'middle',][['pheno.strength']] <- 'middle.weak'
  } else if(middle.value > additive.value) {
    enrich.phenotypes.df[enrich.phenotypes.df$pheno.strength == 'middle',][['pheno.strength']] <- 'middle.strong'
  }
  
  return(enrich.phenotypes.df)
}

##Used in dominanceCalling
dominanceAssign <- function(x, type) {
  if(type == 'BPH') {
    test.dominance <- testForDominanceBPH(x = x)
    test.overdominance <- testForOverdominanceBPH(x = x)
    test.result <- c(test.dominance, test.overdominance)
    stopifnot(!length(grep('ambiguous', test.result)) == 0)
    if(length(grep('ambiguous', test.result)) == 1) {
      x[['dominance']] <- test.result[test.result != 'ambiguous']
    } else { x[['dominance']] <- 'ambiguous' }
    return(x)
    
  } else if(type == 'WPH') {
    test.dominance <- testForDominanceWPH(x = x)
    test.overdominance <- testForOverdominanceWPH(x = x)
    test.result <- c(test.dominance, test.overdominance)
    stopifnot(!length(grep('ambiguous', test.result)) == 0)
    if(length(grep('ambiguous', test.result)) == 1) {
      x[['dominance']] <- test.result[test.result != 'ambiguous']
    } else { x[['dominance']] <- 'ambiguous' }
    return(x)
  }
}

##The following functions are used in dominanceAssign
testForDominanceBPH <- function(x) {
  homo.genotypes <- c('WW', 'NN')
  dominance <- NULL
  if(TRUE %in% (x[x[['status']] == 'enriched',][['genotype']] == homo.genotypes)) {
    enriched.homo <- homo.genotypes[x[x[['status']] == 'enriched',][['genotype']] == homo.genotypes]
    if(length(enriched.homo) == 1) {
      if(x[x[['genotype']] == 'NW',][['status']] != 'depleted') {
        other.homo <- homo.genotypes[homo.genotypes != enriched.homo]
        if(x[x[['genotype']] == enriched.homo,]$pheno.value > x[x[['genotype']] == other.homo,]$pheno.value) {
          dominance <- 'dominant'
        } else { dominance <- 'ambiguous' }
      } else {
        other.homo <- homo.genotypes[homo.genotypes != enriched.homo]
        if(x[x[['genotype']] == enriched.homo,]$pheno.value > x[x[['genotype']] == other.homo,]$pheno.value) {
          dominance <- 'homozygous.genotype.enriched'
        } else{ dominance <- 'ambiguous' }
      }
    } else if(length(enriched.homo) == 2) { dominance <- 'ambiguous' }
  } else { dominance <- 'ambiguous' }
  
  return(dominance)
}

testForOverdominanceBPH <- function(x) {
  dominance <- NULL
  if(length(grep('enriched', x[['status']])) == 1) {
    if(x[x[['status']] == 'enriched',][['genotype']] == 'NW') {
      dominance <- 'overdominant'
    } else { dominance <- 'ambiguous' }
  } else { dominance <- 'ambiguous' }
  
  return(dominance)
}

testForUnderdominanceBPH <- function(x) {
  dominance <- NULL
  if(length(grep('depleted', x[['status']])) == 1) {
    if(x[x[['status']] == 'depleted',][['genotype']] == 'NW') {
      dominance <- 'underdominant'
    } else { dominance <- 'ambiguous' }
  } else { dominance <- 'ambiguous' }
  
  return(dominance)
}

testForDominanceWPH <- function(x) {
  ##Only difference to testForDominanceBPH is that here it is dominant if enriched < not-enriched
  homo.genotypes <- c('WW', 'NN')
  dominance <- NULL
  if(TRUE %in% (x[x[['status']] == 'enriched',][['genotype']] == homo.genotypes)) {
    enriched.homo <- homo.genotypes[x[x[['status']] == 'enriched',][['genotype']] == homo.genotypes]
    if(length(enriched.homo) == 1) {
      if(x[x[['genotype']] == 'NW',][['status']] != 'enriched') {
        other.homo <- homo.genotypes[homo.genotypes != enriched.homo]
        if(x[x[['genotype']] == enriched.homo,]$pheno.value < x[x[['genotype']] == other.homo,]$pheno.value) {
          dominance <- 'dominant'
        } else { dominance <- 'ambiguous' }
      } else {
        other.homo <- homo.genotypes[homo.genotypes != enriched.homo]
        if(x[x[['genotype']] == enriched.homo,]$pheno.value < x[x[['genotype']] == other.homo,]$pheno.value) {
          dominance <- 'homozygous.genotype.enriched'
        } else{ dominance <- 'ambiguous' }
      }
    } else if(length(enriched.homo) == 2) { dominance <- 'ambiguous' }
  } else { dominance <- 'ambiguous' }
  
  return(dominance)
}

testForOverdominanceWPH <- function(x) {
  dominance <- NULL
  if(length(grep('enriched', x[['status']])) == 1) {
    if(x[x[['status']] == 'enriched',][['genotype']] == 'NW') {
      dominance <- 'overdominant'
    } else { dominance <- 'ambiguous' }
  } else { dominance <- 'ambiguous' }
  
  return(dominance)
}

testForUnderdominanceWPH <- function(x) {
  dominance <- NULL
  if(length(grep('depleted', x[['status']])) == 1) {
    if(x[x[['status']] == 'depleted',][['genotype']] == 'NW') {
      dominance <- 'underdominant'
    } else { dominance <- 'ambiguous' }
  } else { dominance <- 'ambiguous' }
  
  return(dominance)
}

######
## This function gives you the fraction of the dominance variables
## heterosis.type = 'BPH' or 'WPH'
## dominance.type = 'dominant', 'overdominant', 'underdominant', 'homozygous.genotype.enriched' or 'ambiguous'
####
dominanceFractionCalling <- function(heterosis.type, dominance.type) {
  dominance.call.df.s <- split(dominance.call.df, dominance.call.df$L2)
  fraction <- lapply(dominance.call.df.s, dominanceFraction,
                     heterosis.type = heterosis.type, dominance.type = dominance.type)
  fraction.m <- melt(fraction)
  names(fraction.m)[2] <- 'alpha.threshold'
  fraction.m[['heterosis.type']] <- heterosis.type
  fraction.m[['dominance.type']] <- dominance.type
  return(fraction.m)
}

dominanceFraction <- function(x, heterosis.type, dominance.type) {
  ##The divisions with 3 is because each marker has 3 genotypes
  ##Doesn't really matter since I divide by 3 for both, but still..
  total.length <- length(grep(heterosis.type, x[['heterosis']])) / 3
  length.of.interest <- length(grep(paste('^', dominance.type, '$', sep = ''),
                                    x[x[['heterosis']] == heterosis.type,]$dominance)) / 3
  fraction <- length.of.interest / total.length
  return(fraction)
}

######
## This function adds the dominance to the homozygous.genotype.enriched to get all cases where the homozygous
## is enriched, regardless of the heterozygous state
####
addDominanceAndHomozygous.genotype.enriched <- function(x) {
  y <- x[x[['dominance.type']] == 'dominant',][['value']] + x[x[['dominance.type']] == 'homozygous.genotype.enriched',][['value']]
  x[x[['dominance.type']] == 'homozygous.genotype.enriched',][['value']] <- y
  
  return(x)
}
  

```

```{r function overdominance in BPH vs entire population}
######
## This function does a t.test to see if QTLs showing overdominance for BPH POLs also
## shows this behavior when looking at the entire populations phenotypes
####
overDominancePhenoDistT.test <- function(x) {
  y <- aggregate(data = x, pheno.value ~ value, mean, na.rm = TRUE)
  bestHomo <- y[y$pheno.value == max(y[y$value == 'NN',]$pheno.value,
                                    y[y$value == 'WW',]$pheno.value),]$value
  tTest <- t.test(x = x[x$value == bestHomo,]$pheno.value,
                  y = x[x$value == 'NW',]$pheno.value)
  k <- NULL
  if(y[y$value == bestHomo,]$pheno.value > y[y$value == 'NW',]$pheno.value) {
    k <- 'homoBest'
  } else {
    k <- 'hetBest'
  }
  effect.size <- dist(y[y$value == bestHomo,]$pheno.value,
                      y[y$value == 'NW',]$pheno.value)
  df <- data.frame(p.value = tTest$p.value, status = k, effectSize = effect.size)
  return(df)
}
```

```{r function amount of genotypes in heterosis classes}
######
## This function calculats the amount of genotypes in the entire population,
## and in the BPH and WPH fraction of the population. This is for the paper
## where I look at overdominance and dominance in BPH hybrids
## Figure 4 in the paper. I do this on the fly for now
####
genotypeHeterosis <- function(phenotype, marker) {
  id.and.marker <- PhenotypeMarkerDistribution(phenotype = phenotype, marker = marker,
                                               output = 'geno.pheno')
  
  het <- data.frame(id = heterosis.data[heterosis.data$L2 == phenotype,]$L3,
                       heterosis = heterosis.data[heterosis.data$L2 == phenotype,]$value)
  id.and.marker.het <- join(het, id.and.marker, by = 'id')
  
  agg.full.data <- aggregate(data = id.and.marker.het, id ~ get(names(id.and.marker.het[3])), length)
  agg <- aggregate(data = id.and.marker.het, id ~ get(names(id.and.marker.het[3])) + heterosis, length)
  agg.out <- agg[grep('BPH|WPH', agg$heterosis),]
  
  out <- data.frame(marker = c(as.character(agg.out[,1]), as.character(agg.full.data[,1])),
                    heterosis = c(as.character(agg.out$heterosis), rep('Entire', 3)), amount = c(agg.out$id, agg.full.data$id))
  
  return(out)
}
```

```{r function significant marker}
#output = plot, df, or highlight
#Important function that I use for pretty much everything... Should break it apart..
sigMark <- function(phenotypeName, output, marker, highlight_df) { #highlight = NULL, if not used
  pheno.value <- original.median[original.median$phenotype == phenotypeName,] #These are the original phenotype values, not the residuals
  pheno.value <- data.frame(id = pheno.value$strain, pheno.value = pheno.value$value)
  if(is.na(marker) == TRUE) {
  markerNames <- total.peaks[total.peaks[['phenotype']] == phenotypeName,][['marker']]
  markerDataset <- total.peaks[total.peaks[['phenotype']] == phenotypeName,][['dataset']]
  markerNames <- paste(markerNames, markerDataset, sep = '_')
  } else { #If markers are designated, as in the heterosisVsDominance function
    markerNames <- marker[marker[['phenotype']] == phenotypeName,][['marker']]
    markerDataset <- marker[marker[['phenotype']] == phenotypeName,][['dataset']]
    markerNames <- paste(markerNames, markerDataset, sep = '_')
  }
  
  if(length(markerNames) >= 1) {
    grepMarkers <- paste('id', paste(gsub('?_.*$', '', markerNames), collapse = '|'), sep = '|')
    genoValue <- geno[,grep(grepMarkers, names(geno))]
    genoPheno <- merge(genoValue, pheno.value, by = 'id')
    genoPheno.m <- na.exclude(melt(genoPheno, measure.vars = gsub('?_.*$', '', markerNames)))
    genoPheno.s <- split(genoPheno.m, genoPheno.m$variable)
    for(i in 1:length(genoPheno.s)) {
      genoPheno.s[[i]]$dataset <- markerDataset[i]
    } #Adds column for each list object data frame with if it's parental or residual
    genoPheno.m <- do.call(rbind, genoPheno.s)

    markerPhenoPlot <- ggplot(genoPheno.m, aes(value, pheno.value, fill = value, color = dataset)) +
      scale_fill_manual(values=cb.palette) + geom_boxplot() + theme_bw() +
      facet_wrap(~variable, ncol = 4) +
      labs(y = phenotypeName, x = NULL, title = phenotypeName) +
      theme(legend.position = 'none') +
      theme(axis.title.x = element_text(size = sizeAxisTitle, family = fontFamily)) +
      theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) + 
      theme(axis.text.x = element_text(size = sizeAxisLabels, family = fontFamily)) +
      theme(axis.text.y = element_text(size = sizeAxisLabels, family = fontFamily)) +
      theme(strip.text.x = element_text(size = sizeFacetTitles, family = fontFamily))

    if(length(highlight_df$id) >= 1) { #check heterosis coupled with dominance
      highlight.s <- split(highlight_df, f = highlight_df$heterosis)
      BPH <- highlight.s$BPH
      WPH <- highlight.s$WPH
      
      ##Strains with missing genotype at the marker or missing phenotype will not be caught here
      BPHvalues <- genoPheno.m[grep(paste(paste('^', BPH$id, '$', sep =''), collapse = '|'), genoPheno.m$id),]
      WPHvalues <- genoPheno.m[grep(paste(paste('^', WPH$id, '$', sep =''), collapse = '|'), genoPheno.m$id),]

      BPHvalues$heterosis <- 'BPH'
      WPHvalues$heterosis <- 'WPH'

      highlight <- rbind(BPHvalues, WPHvalues)

      markerPhenoPlot <- markerPhenoPlot +
        geom_jitter(data = highlight, aes(value, pheno.value), shape = 16, alpha = 0.25) + 
        facet_grid(heterosis ~ variable) +
        theme(axis.title.x = element_text(size = sizeAxisTitle, family = fontFamily)) +
        theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) + 
        theme(axis.text.x = element_text(size = sizeAxisLabels, family = fontFamily)) +
        theme(axis.text.y = element_text(size = sizeAxisLabels, family = fontFamily)) +
        theme(strip.text.x = element_text(size = sizeFacetTitles, family = fontFamily)) +
        theme(strip.text.y = element_text(size = sizeFacetTitles, family = fontFamily))

      highlight.comparison.s <- split(highlight, list(highlight$variable, highlight$heterosis,
                                                      highlight$dataset))

      highlight.comparison.results <- list()
      for(i in names(highlight.comparison.s)) {
        highlight.comparison.results[[i]] <- highlightComparison(x = highlight.comparison.s[[i]])
      }

      highlight.comp.list.all <- do.call(rbind, highlight.comparison.results)
      highlight.comp.df <- do.call(rbind, highlight.comp.list.all[,1])
      highlight.comp.df.chi <- do.call(rbind, highlight.comp.list.all[,2])

      highlight.comp.df$marker <- gsub('?\\..*$', '', row.names(highlight.comp.df))
      highlight.comp.df$genotype <- gsub('^.*?\\.', '', row.names(highlight.comp.df))
      highlight.comp.df$heterosis <- gsub('?\\..*$', '', highlight.comp.df$genotype)
      highlight.comp.df$genotype <- gsub('^.*?\\.', '', highlight.comp.df$genotype)
      highlight.comp.df$dataset <- gsub('?\\..*$', '', highlight.comp.df$genotype)
      highlight.comp.df$genotype <- gsub('^.*?\\.', '', highlight.comp.df$genotype)
      highlight.comp.df <- highlight.comp.df[highlight.comp.df$heterosisStrains != 'NaN',]
      highlight.comp.df.s <- split(highlight.comp.df, f = highlight.comp.df$marker)

      highlight.comp.df.chi$marker <- gsub('?\\..*$', '', row.names(highlight.comp.df.chi))
      highlight.comp.df.chi$genotype <- gsub('^.*?\\.', '', row.names(highlight.comp.df.chi))
      highlight.comp.df.chi$heterosis <- gsub('?\\..*$', '', highlight.comp.df.chi$genotype)
      highlight.comp.df.chi$genotype <- gsub('^.*?\\.', '', highlight.comp.df.chi$genotype)
      highlight.comp.df.chi$dataset <- gsub('?\\..*$', '', highlight.comp.df.chi$genotype)
      highlight.comp.df.chi$genotype <- gsub('^.*?\\.', '', highlight.comp.df.chi$genotype)
      highlight.comp.df.chi <- highlight.comp.df.chi[highlight.comp.df.chi$expected != 'NaN',]
      highlight.comp.df.chi <- split(highlight.comp.df.chi, f = highlight.comp.df.chi$marker)

      PrepHet <- function(x) {
        y <- split(x, x$heterosis)
        for(i in names(y)) names(y[[i]])[1] <- i
        y$WPH <- y$WPH[,-c(2,3,4,5)]
        y$BPH <- y$BPH[,-5]

        z <- do.call(cbind, y)
        names(z) <- gsub('BPH.', '', names(z))
        return(z)
      }

      highlight.comp.df <- lapply(X = highlight.comp.df.s, FUN = PrepHet)

      highlight.comp.df.m <- melt(highlight.comp.df)
      highlight.comp.df.m$marker <- factor(highlight.comp.df.m$marker,levels = 
                                             unique(highlight.comp.df.m$marker))
      highlight.comp.df.m$variable <- gsub('WPH.WPH', 'WPH', highlight.comp.df.m$variable)
      
      highlightVsOriginal <- ggplot(highlight.comp.df.m, aes(variable, value,
                                                               fill = genotype)) +
        geom_bar(stat = 'identity') + scale_fill_manual(values=cb.palette) +
        facet_wrap(~marker + dataset, nrow = 1) +
        labs(x = NULL, y = NULL, title = paste(phenotypeName, sep = ' ')) +
        theme(axis.title.x = element_text(size = sizeAxisTitle, family = fontFamily)) +
        theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) + 
        theme(axis.text.x = element_text(size = sizeAxisLabels, family = fontFamily)) +
        theme(axis.text.y = element_text(size = sizeAxisLabels, family = fontFamily)) +
        theme(strip.text.x = element_text(size = sizeFacetTitles, family = fontFamily),
            strip.text.y = element_text(size = sizeFacetTitles, family = fontFamily)) +
        theme(legend.title = element_blank()) +
        theme(legend.text = element_text(size = sizeLegend, family = fontFamily))

      high.output <- list(plot = markerPhenoPlot, comparisonDf = highlight.comp.df,
                            comparisonDFchi = highlight.comp.df.chi, comparisonPlot = highlightVsOriginal)

    }

    aggregate.list <- list()
    for(i in gsub('?_.*$', '', markerNames)) {
        aggregate.list[[i]] <- aggregate(data = genoPheno, pheno.value ~ get(i), FUN = median)
        aggregate.list[[i]]$marker <- i
    }
    for(i in 1:length(aggregate.list)) {
      aggregate.list[[i]]$dataset <- markerDataset[i]
    }
    df <- do.call(rbind, aggregate.list)
    
    #hack for the galactose were the chr4 qtl has the same pos in res and par...
    if(length(grep('galactose_GT', phenotypeName) == 1)) {
      parchr4gal <- aggregate.list$`chr4-463889`
      parchr4gal$dataset <- 'parental'
      row.names(parchr4gal) <- row.names(df[df$marker == 'chr4-463889',])
      df <- rbind(df, parchr4gal)
    }
    names(df)[1] <- 'genotype'
    df$phenotype <- phenotypeName

    if(output == 'plot') {
      return(markerPhenoPlot) } else if(output == 'df') {
      return(df) } else if(output == 'highlight') {
        return(high.output) } else if(output == 'genoPheno') {
          return(genoPheno) }
  }
}

##Function for calculating frequencies in heterosis and dominance comparison, used in sigMark
highlightComparison <- function(x, env = parent.frame()) {
  marker <- as.character(unique(x$variable))
  original <- env$genoPheno.m[env$genoPheno.m$variable == marker,]
  high.length.NN <- length(x[x$value == 'NN',]$id)
  high.length.NW <- length(x[x$value == 'NW',]$id)
  high.length.WW <- length(x[x$value == 'WW',]$id)
  
  high.NN <- high.length.NN / length(x$id)
  high.NW <- high.length.NW / length(x$id)
  high.WW <- high.length.WW / length(x$id)
  
  original.length.NN <- length(original[original$value == 'NN',]$id)
  original.length.NW <- length(original[original$value == 'NW',]$id)
  original.length.WW <- length(original[original$value == 'WW',]$id)
  
  original.NN <- original.length.NN / length(original$id)
  original.NW <- original.length.NW / length(original$id)
  original.WW <- original.length.WW / length(original$id)
  
  comparison.df.prop <- data.frame(heterosisStrains = c(high.WW, high.NW, high.NN),
                                   original = c(original.WW, original.NW, original.NN))
  row.names(comparison.df.prop) <- c('WW', 'NW', 'NN')
  comparison.df.chi <- data.frame(heterosisStrains = c(high.length.WW, high.length.NW, high.length.NN),
                                  expected = c(original.WW * length(x$id), original.NW * length(x$id),
                                               original.NN * length(x$id)))
  row.names(comparison.df.chi) <- c('WW', 'NW', 'NN')
  comparison <- list(proportion = comparison.df.prop, chi = comparison.df.chi)
  
  return(comparison)
}

```

-----

```{r F12 segregant load data}
load('~/Dropbox/myResearch/My_10k/10kparents_phenotyping/qtl-mapping-F12segregants/F12segregants_qtlscan_151027_clean.Rdata')
F12seg.threshold <- summary(F12seg.cross_OneDimScan_permutations, alpha = alpha.value.qtl)
```

```{r hybrids load analysis}
##This has the genotypes for all possible crosses, not just the ones used in the mapping
load(file = '~/Dropbox/myResearch/My_10k/R_data/all-hybrid-genotypes-160130.Rdata') #hybrid.genotypes

##Read in the phased genotypes for looking at parent-of-origin effects, also contains all possible crosses
##Missing genotypes are removed
load(file = '~/Dropbox/myResearch/My_10k/R_data/phased-genotypes160624.Rdata') #phased.hybrid.genotypes

##Reading in the results of the QTL analysis from the server
##Ready for use in the scripts
load('~/Dropbox/myResearch/My_10k/10khybrids_phenotyping/qtl-mapping-residuals-10khybrids/hybrid_QTL_analysis_clean.Rdata')
load('~/Dropbox/myResearch/My_10k/10khybrids_phenotyping/qtl-mapping-parentals-10khybrids/parent_hybrid_QTL_analysis.Rdata')
#Transforming parent phenotypes so that positive values are good for GT and yield
parent_pheno[,grep('GT', names(parent_pheno))] <- - parent_pheno[,grep('GT', names(parent_pheno))]

##Reading in the data gotten from Kaspar for the LMM qtl mapping, both additive and dominance
lmm <- read.csv('~/Dropbox/myResearch/My_10k/LMM/QTLs_hybrids.csv')
lmm.df <- apply(lmm[,2:19], 2, function(x) -log10(x))
lmm.df <- data.frame(chr = lmm$chrom, pos = lmm$pos, lmm.df)
attr(lmm.df, "class") <- c("scanone", "data.frame")

load('~/Dropbox/myResearch/My_10k/LMM/dominanceLMM/dominance_pvalues.RData')
lmm.dominance <- dominance_pv
lmm.dominance.df <- apply(lmm.dominance[,1:18], 2, function(x) -log10(x))
lmm.dominance.df <- data.frame(chr = lmm$chrom, pos = lmm$pos, lmm.dominance.df)
attr(lmm.dominance.df, "class") <- c("scanone", "data.frame")

##Loading the actual normalised phenotypes, not the residuals used for the mapping. The data is
##transformed so that a high value is good both for GT and Yield
load('~/Dropbox/myResearch/My_10k/R_data/hybrid_norm_median_values.df')
```

```{r formating phenotype data}
##Formating the original hybrid phenotypes
original.median <- hybrid.original.median #changing name
rm(hybrid.original.median) #remove the old object
original.median.all.strains <- original.median #creating a df before removing aneuploid hybrids
original.median.all.strains$variable <- gsub('GenerationTime', 'GT', original.median.all.strains$variable)
original.median.all.strains$phenotype <- paste(original.median.all.strains$environment,
                                               original.median.all.strains$variable, sep = '_')

##Removing all strains that were not used in the qtl mapping 
##Main reason for less strains for the mapping is our decision to remove the ones with chr9 aneuploidy
rm.strains.not.in.qtl <- setdiff(original.median$strain, pheno$id)
rm.strains.not.in.qtl.grep <- paste('^', rm.strains.not.in.qtl, '$', sep = '', collapse = '|')
original.median <- original.median[-grep(rm.strains.not.in.qtl.grep, original.median$strain),]
original.median$variable <- gsub('GenerationTime', 'GT', original.median$variable) #Needs to be the same names as the QTL data
original.median$phenotype <- paste(original.median$environment, original.median$variable, sep = '_')
```

```{r aneuploid strains vs entire population}
##Calculating significant differences between chr IX anueploid strain distribution and the rest
aneuploid.strains <- '^206|^222|^223|^254|^258|41$|53$|67$'
original.median.all.strains.s <- split(original.median.all.strains, original.median.all.strains$phenotype)
aneuplody.ttest <- lapply(original.median.all.strains.s, FUN = function(x) t.test(x[grep(aneuploid.strains, x[['strain']]),]$value,
                                                                                  x[-grep(aneuploid.strains, x[['strain']]),]$value))
aneuplody.ttest.summary <- melt(lapply(aneuplody.ttest, FUN = function(x) x[[3]] < 0.01))
##Most of the distributions differ significantly, I'll plot all of them in supplementary figure 1 in the paper

##Creating the plot comparing the aneuploid strains with the rest
##This plot is used in the supplementary information
plot.anueploid.distribution.vs.rest <- ggplot(original.median, aes(environment, value, fill = environment)) +
    scale_fill_manual(values = c("#000000", cb.palette)) + geom_violin(alpha = 0.5) + 
    geom_violin(data = original.median.all.strains[grep(aneuploid.strains, original.median.all.strains$strain),],
                aes(environment, value), alpha = 0.5, fill = 'red') +
    facet_wrap(~variable) + theme_bw() + theme(legend.position = 'none')
#svg(filename = '~/Dropbox/myResearch/My_10k/figures/finals/supplementary/chrIXaneuploidy-distributions.svg', width = 8, height = 3)
```

```{r phenotype data for supplementary data in paper}
##Putting the phenotype data in order for use in the supplementary data for the paper
##Using original.median.all.strains for the hybrids and the appropriate phenotype data from F12seg.crossdata
original.median.paper <- original.median.all.strains
original.median.paper$type <- 'POL'
original.median.paper$variable[original.median.paper$variable == 'Yield'] <- 'Total_growth'
original.median.paper$variable[original.median.paper$variable == 'GT'] <- 'Growth_rate'
original.median.paper <- original.median.paper[names(original.median.paper) != 'phenotype']

F12seg.dataPaper <- F12seg.crossdata$pheno
F12seg.dataPaper2 <- F12seg.dataPaper[,-grep('ura|CuCl|LiCl|arsenite', names(F12seg.dataPaper))]
names(F12seg.dataPaper2) <- c('strain', 'caffeine', 'hydroxyurea', 'NaCl', 'galactose', 'allantoin',
                              'glycine', 'isoleucine', 'phleomycin', 'rapamycin')
F12seg.dataPaper3 <- melt(F12seg.dataPaper2,  id.vars = 'strain')
names(F12seg.dataPaper3)[2] <- 'environment'
F12seg.dataPaper3$variable <- 'Growth_rate'
F12seg.dataPaper3$type <- 'F12_haploid_parent'
F12seg.dataPaper4 <- F12seg.dataPaper3[c(4, 2, 1, 5, 3)]

phenotype.data.for.paper <- rbind(original.median.paper, F12seg.dataPaper4)
names(phenotype.data.for.paper)[1] <- 'growth_measure'
names(phenotype.data.for.paper)[5] <- 'normalized_value'
```

```{r data variables}
threshold <- summary(cross_OneDimScan_permutations, alpha = alpha.value.qtl)
parent.threshold <- summary(parent_cross_OneDimScan_permutations, alpha = alpha.value.qtl)
pheno.names <- names(pheno)[-1] #Names of the phenotypes
environments <- unique(gsub('?_.*$', '', pheno.names)) #Name of the environments
geno <- geno[-c(1,2),]
names(geno) <- gsub('\\.', '-', names(geno))
```

```{r plot POLs F12 parents estimated parents comparison}
######
## This chunk looks into the phenotypic differences between the hybrids, estimated parents and haploid parents
####
##Put the data in to a format compatible with the plot making
F12.vs.hybrid.gt <- list()
for(env in unique(original.median$environment)) {
  F12.vs.hybrid.gt[[env]] <- phenoDistCompF12segHybrid(environment = env, pheno = 'generationTime')
}

F12.vs.hybrid.yield <- list()
for(env in unique(original.median$environment)) {
  F12.vs.hybrid.yield[[env]] <- phenoDistCompF12segHybrid(environment = env, pheno = 'yield')
}

F12.vs.hybrid.gt <- do.call(rbind, F12.vs.hybrid.gt)
F12.vs.hybrid.yield <- do.call(rbind, F12.vs.hybrid.yield)

######
## These plots are in the paper, this one in Figure 1 and the other in a supplementary figure
####
plot.F12.vs.hybrid.gt <- ggplot(F12.vs.hybrid.gt, aes(dataset, value, fill = dataset)) + geom_violin() +
  scale_fill_manual(values=cb.palette) +
  geom_violin(data = F12.vs.hybrid.gt[F12.vs.hybrid.gt$dataset == 'F12 segregants',], aes(fill = ura), alpha = 0.5) +
  geom_violin(data = F12.vs.hybrid.gt[F12.vs.hybrid.gt$dataset == 'Estimated Parents',], aes(fill = ura), alpha = 0.5) +
  facet_wrap(~environment) + labs(x = NULL, y = 'Normalized doubling time') +
  theme_bw() + theme(panel.grid.minor = element_blank()) +
  theme(axis.text.x = element_text(size = sizeAxisLabels, family = fontFamily, angle = axisAnglesX,
                                   hjust = 1)) +
  theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) +
  theme(axis.text.y = element_text(size = sizeAxisLabels, family = fontFamily)) +
  theme(strip.text.x = element_text(size = sizeFacetTitles, family = fontFamily)) +
  theme(legend.text = element_text(size = sizeLegend, family = fontFamily)) +
  theme(legend.title = element_blank()) +
  scale_x_discrete(labels = c('F12 parents', 'Hybrids', 'Estimated parents'))
#svg(filename = '~/Dropbox/myResearch/My_10k/figures/finals/phenotypeDistRate.svg', width = 10, height = 8)

plot.F12.vs.hybrid.yield <- ggplot(F12.vs.hybrid.yield, aes(dataset, value, fill = dataset)) + geom_violin() +
  scale_fill_manual(values=cb.palette) +
  geom_violin(data = F12.vs.hybrid.yield[F12.vs.hybrid.yield$dataset == 'Estimated Parents',], aes(fill = ura),
              alpha = 0.5) + facet_wrap(~environment) +
  theme_bw() + theme(panel.grid.minor = element_blank()) +
  labs(x = NULL, y = 'Normalized total growth') +
  theme(axis.text.x = element_text(size = sizeAxisLabels, family = fontFamily, angle = axisAnglesX,
                                   hjust = 1)) +
  theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) +
  theme(axis.text.y = element_text(size = sizeAxisLabels, family = fontFamily)) +
  theme(strip.text.x = element_text(size = sizeFacetTitles, family = fontFamily)) +
  theme(legend.text = element_text(size = sizeLegend, family = fontFamily)) +
  theme(legend.title = element_blank())
```

```{r plotting QTL maps for main text}
##Just checked, and these aren't the plots I used for the paper.. 160802
QTL.maps <- list()
for(i in environments) {
  QTL.maps[[i]] <- HomeMadeQtlMap(environment = i, parent_or_residual = 'residual')
}
```

```{r plot QTLs for supplementary}
######
## These plots show each environment with both generation time (in orange) and yield (in blue)  
## Both for the hybrid residuals and the estimated parental phenotypes  
## The dashed lines represent a `r alphaValue` threshold and have matching colors to corresponding phenotype
## Plotting the QTL maps used as supplementary figures
####
#for(i in environments) {
#  pltQTL(env = i)
#}
```

```{r extract QTL peaks}
##From residual scan
peaks.res <- list()
for (i in pheno.names) {
    peaks.res[[i]] <- extractPeaks(phenotype = i, dataSet = 'residual')
}

residual.peaks <- peakFinalize(peaks = peaks.res, n = 6642)

##From parental scan
peaks.par <- list()
for (i in pheno.names) {
    peaks.par[[i]] <- extractPeaks(phenotype = i, dataSet = 'parent')
}

parent.peaks <- peakFinalize(peaks = peaks.par, n = 164)
```

```{r extract multiple QTLs on chromosomes}
##Picking out QTLs that were on the same chromosome as a higher peak, and thus not recognized by R/qtl
multiple.qtls <- list()
for(i in pheno.names) {
  multiple.qtls[[i]] <- MultiQTL(phenotype = i)
}

add.qtls <- list(glycine_GT = '13.612702', isoleucine_GT = '13.410894',
                 hydroxyurea_Yield = '7.372503', phleomycin_Yield = '9.26336',
                 phleomycin_Yield = '15.425274', allantoin_GT = '13.31800',
                 allantoin_GTT = '13.420348', caffeine_GT = '9.73949',
                  caffeine_Yield = '9.71330')

add.qtls.for.finalize <- list()
for(names in names(add.qtls)) {
  add.qtls.for.finalize[names] <- addQTLS(x = add.qtls[names])
}

add.peaks <- peakFinalize(peaks = add.qtls.for.finalize, n = 6642)
add.peaks$marker <- paste('chr', add.peaks$marker, sep = '')
add.peaks$intLength <- '-'

```

```{r QTL peaks summary}
##Putting together the peaks into one data frame, I'll use this for the supplementary information and for 
total.peaks <- rbind(residual.peaks, add.peaks, parent.peaks)

##This figure is in the main text, it shows how much variation the QTLs explained, split according to additive and nonadditive
plot.individual.qtl.variance.explained <- ggplot(total.peaks, aes(dataset, VarianceExplained, fill = dataset)) +
  geom_boxplot() + theme_bw()

##T.test to check for difference between additive and nonadditive QTLs power to explain variation
total.peaks.variance.explained.t.test <- t.test(total.peaks[total.peaks$dataset == 'residual',]$VarianceExplained,
                                              total.peaks[total.peaks$dataset == 'parental',]$VarianceExplained, paired = FALSE, var.equal = FALSE)

##Mean variation explained by QTLs, used in the heterosis section in the paper
explained.variance.difference <- aggregate(data = total.peaks, VarianceExplained ~ dataset,
                                         mean, na.rm = T)

##Formating the peaks used for the supplementary information and for the variance explained chunk below
total.peaks.paper <- total.peaks
total.peaks.paper$dataset[total.peaks.paper$dataset == 'residual'] <- 'Non-additive'
total.peaks.paper$dataset[total.peaks.paper$dataset == 'parental'] <- 'Additive'
total.peaks.paper$phenotype <- as.character(total.peaks.paper$phenotype)
total.peaks.paper$marker <- as.character(total.peaks.paper$marker)
#print(total.peaks.paper, row.names = FALSE)
##Put -'s on the lod interval for the added qtls. Can't calculate an interval for them.
```

```{r variance explained by QTLs per environment}
#Using total.peaks.paper for this
variance.explained.qtl <- aggregate(data = total.peaks.paper, VarianceExplained ~ phenotype + dataset, sum)

##This is used as a supplementary figure showing how much of the variance in a phenotype the found QTLs explain
plot.environment.qtl.variance.explained <- ggplot(variance.explained.qtl, aes(phenotype, VarianceExplained, fill = dataset)) +
  facet_wrap(~dataset, ncol = 1) + geom_bar(stat = 'identity') + theme_bw() + theme(legend.position = 'none')
#svg(filename = '~/Dropbox/myResearch/My_10k/figures/review-NatureCom/variance-explained-qtls.svg', width = 7, height = 4)
```

```{r number of pleiotropic qtls}
##I get these numbers from the QTL summary Excel
pleiotropic.qtls <- data.frame(bins = c(2,3,4,8,9), qtls = c(4,1,0,3,1))

##This plot is used in the paper in the QTL figure to show the pleiotropic QTLs
plot.pleiotropic.qtls <- ggplot(pleiotropic.qtls, aes(as.factor(bins), qtls)) + geom_bar(stat = 'identity') + theme_bw()

#svg(filename = '~/Dropbox/myResearch/My_10k/figures/finals/pleiotropyDist.svg', width = 5, height = 4)
```

```{r load data for heterosis calculations}
##Loading the phenotypes where I also have the replicates in order to get a variance 
##This data is already transformed so that high GT is good, just as in Yield
##This data was constructed in the final_phenotypeAnalysisDistribution scripts
load('~/Dropbox/myResearch/My_10k/R_data/raw_hybrid_data.Rdata') #hybrid.raw.clean

##Picking out the information I want from hybrid.raw.clean, removing aneuploid strains
hybrid.raw.clean.2way <- hybrid.raw.clean[hybrid.raw.clean$type == '2way-2way',]
hybrid.raw.clean.2way$variable <- gsub('GenerationTime', 'GT', hybrid.raw.clean.2way$variable)
hybrid.raw.clean.2way$phenotype <- paste(hybrid.raw.clean.2way$environment,
                                         hybrid.raw.clean.2way$variable, sep = '_')
hybrid.raw.clean.2way <- hybrid.raw.clean.2way[-grep('^206|^222|^223|^254|^258|41$|53$|67$', hybrid.raw.clean.2way$strain),]

##Calculating the SE here with the original data with all the replicates
hybrid.SE <- aggregate(data = hybrid.raw.clean.2way, value ~ environment + variable + strain, FUN = SE)
mean.hybrid.SE <- mean(hybrid.SE$value, na.rm = TRUE)
```

```{r heterosis calculations}
##Calculating heterosis for the range of alpha values
heterosis.alpha.values.list <- as.list(heterosis.alpha.values)

heterosis.data.alpha.comparison <- pblapply(heterosis.alpha.values.list, FUN = heterosisAnalysis, centralTendency = mean)

heterosis.data.alpha.comparison.m <- melt(heterosis.data.alpha.comparison)
heterosis.data.alpha.comparison.m.s <- split(heterosis.data.alpha.comparison.m, heterosis.data.alpha.comparison.m$L1)

##Alpha value comparison
heterosis.data.alpha.comparison.results <- pblapply(heterosis.data.alpha.comparison.m.s, heterosisDataAlphaComparisonSummary)
heterosis.data.alpha.comparison.results <- melt(heterosis.data.alpha.comparison.results)
heterosis.data.alpha.comparison.results$pValue <- as.numeric(gsub('p', '', heterosis.data.alpha.comparison.results$pValue))
heterosis.data.alpha.comparison.resultsForPlot <- heterosis.data.alpha.comparison.results[-grep('AsMP', heterosis.data.alpha.comparison.results$variable),]

######
## This shows the fraction of all the different heterosis types as a function of the pvalues used
## Use for the following section in the paper ##heterosis.data.alpha.comparison.results.s
  #"the majority (83% to 85%) deviated significantly from the midparent expectation and were thus heterotic. Half (55 to 65%) 
  #of these cases corresponded to the offspring being either superior (best parent heterosis) or inferior (worst parent heterosis)
  #to both parents, with equal prevalence of best parent and worst parent heterosis (Fig. 4a)."
## AsMP, PosMPH and NegMPH adds up to 1. BPH is a fraction of the PosMPH while WPH is a fraction of NegMPH
## All these are fractions of 1-nonSignificant
####
heterosis.data.alpha.comparison.results.s <- split(heterosis.data.alpha.comparison.results,
                                               heterosis.data.alpha.comparison.results$pValue)
##Calculating values for the paper
heterosis.data.alpha.comparison.results.s.m <- do.call(rbind, heterosis.data.alpha.comparison.results.s)
heterosis.max.amount.of.heterotic <- 1 - min(heterosis.data.alpha.comparison.results.s.m[heterosis.data.alpha.comparison.results.s.m$variable =='AsMP',]$value)
heterosis.min.amount.of.heterotic <- 1 - max(heterosis.data.alpha.comparison.results.s.m[heterosis.data.alpha.comparison.results.s.m$variable =='AsMP',]$value)

heterosis.BPH.or.WPH.max <- max(do.call(rbind, lapply(heterosis.data.alpha.comparison.results.s,
                                                 FUN = function(x) x[x$variable == 'BPH',]$value + x[x$variable == 'WPH',]$value)))
heterosis.BPH.or.WPH.min <- min(do.call(rbind, lapply(heterosis.data.alpha.comparison.results.s,
                                                 FUN = function(x) x[x$variable == 'BPH',]$value + x[x$variable == 'WPH',]$value)))



##Picking out the predefined alpha value
heterosis.data <- heterosis.data.alpha.comparison.m.s[[paste('p', heterosis.alpha.value, sep = '')]] #Used for dominance calculations
heterosis.data.summary <- heterosis.data.alpha.comparison.results.s[[as.character(heterosis.alpha.value)]]
```

```{r plot heterosis}
##This plot is used for figure 4 in the paper and shows the fraction of heterotic strains
ggplot(heterosis.data.alpha.comparison.resultsForPlot[heterosis.data.alpha.comparison.resultsForPlot$variable != 'nonSignificant',],
       aes(pValue, value, color = variable, group = variable)) +
  scale_color_manual(values = cb.palette[2:8]) + geom_line(size = 1.5) + geom_point(size = 2.5) +
  labs(x = 'FDR', y = 'Frequency') +
  theme_bw() + theme(panel.grid.minor = element_blank()) +
  theme(axis.text.x = element_text(size = sizeAxisLabels,
                                   family = fontFamily, angle = axisAnglesX, hjust = 1)) +
  theme(axis.title.x = element_text(size = sizeAxisTitle, family = fontFamily)) +
  theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) +
  theme(axis.text.y= element_text(size = sizeAxisLabels, family = fontFamily)) +
  theme(strip.text.x = element_text(size = sizeFacetTitles, family = fontFamily)) +
  theme(legend.text = element_text(size = sizeLegend, family = fontFamily)) +
  theme(legend.title = element_blank()) +
  scale_y_continuous(limits = c(0, 1))  + coord_trans(x = 'log2') +
  scale_x_continuous(breaks = as.numeric(heterosis.alpha.values)) +
  theme(legend.position = 'none')

#svg(filename = '~/Dropbox/myResearch/My_10k/figures/review-NatureCom/heterosisFrequency.svg',
#    width = 4.47, height = 1.97)

##This plot is used for a supplementary figure in the paper and shows the fraction of non 
##significantly different hybrids from the parents
ggplot(heterosis.data.alpha.comparison.resultsForPlot[heterosis.data.alpha.comparison.resultsForPlot$variable == 'nonSignificant',],
       aes(pValue, value, color = variable, group = variable)) +
  scale_color_manual(values = cb.palette[1]) + geom_line(size = 1.5) + geom_point(size = 2.5) +
  labs(x = 'FDR', y = 'Frequency') +
  theme_bw() + theme(panel.grid.minor = element_blank()) +
  theme(axis.text.x = element_text(size = sizeAxisLabels, family = fontFamily, angle = axisAnglesX, hjust = 1)) +
  theme(axis.title.x = element_text(size = sizeAxisTitle, family = fontFamily)) +
  theme(axis.title.y = element_text(size = sizeAxisTitle, family = fontFamily)) +
  theme(axis.text.y= element_text(size = sizeAxisLabels, family = fontFamily)) +
  theme(strip.text.x = element_text(size = sizeFacetTitles, family = fontFamily)) +
  theme(legend.text = element_text(size = sizeLegend, family = fontFamily)) +
  theme(legend.title = element_blank()) +
  scale_y_continuous(limits = c(0.1, 1)) + coord_trans(x = 'log2') +
  scale_x_continuous(breaks = as.numeric(heterosis.alpha.values)) +
  theme(legend.position = 'none')
#svg(filename = '~/Dropbox/myResearch/My_10k/figures/review-NatureCom/heterosisFrequencyNotSignificant.svg',
#    width = 8, height = 2)
```

```{r dominance calculations}
##Here I create the heterosis.vs.dominance list. It's very important for the overdominance and dominance calculations
heterosis.vs.dominance <- list()
for(i in pheno.names) {
  print(i)
  heterosis.vs.dominance[[i]] <- HeterosisVsDominance(phenotype = i)
}

##I do the chi square test for all environments and QTLs
heterosis.vs.dominance.chiSq <- list()
for (i in names(heterosis.vs.dominance)) {
  heterosis.vs.dominance.chiSq[[i]] <- lapply(heterosis.vs.dominance[[i]][['comparisonDFchi']], chiSquareTest)
}

##Putting the heterosis.vs.dominance.chiSq in a better format to use below
chiSq.df <- list()
for(i in names(heterosis.vs.dominance.chiSq)) {
  chiSq.df[[i]] <- do.call(rbind, heterosis.vs.dominance.chiSq[[i]])
}
chiSq.df <- do.call(rbind, chiSq.df)
chiSq.df$pheno <- gsub('?\\..*$', '', row.names(chiSq.df))

##Calling dominance
dominance.call <- vector('list', length(chiSq.df$WW))
for(i in 1:length(chiSq.df$WW)) {
  cat(paste('percent done', round(i / length(chiSq.df$WW), 3) * 100, sep = ' -> '))
  for(j in names(dominance.alpha.values)) {
    dominance.call[[i]][[j]] <- dominanceCalling(x = chiSq.df[i,], alpha = dominance.alpha.values[[j]])
  }
}

dominance.call.df <- melt(dominance.call, id.vars = c(names(dominance.call[[1]][[1]])))

ggplot(dominance.call.df, aes(dominance, fill = dominance)) +
  geom_bar() + facet_wrap(~heterosis)

ambiguous.BPH.fractions <- dominanceFractionCalling(heterosis.type = 'BPH', dominance.type = 'ambiguous')
dominant.BPH.fractions <- dominanceFractionCalling(heterosis.type = 'BPH', dominance.type = 'dominant')
homozygous.enrichment.BPH.fractions <- dominanceFractionCalling(heterosis.type = 'BPH', dominance.type = 'homozygous.genotype.enriched')
overdominant.BPH.fractions <- dominanceFractionCalling(heterosis.type = 'BPH', dominance.type = 'overdominant')
underdominant.BPH.fractions <- dominanceFractionCalling(heterosis.type = 'BPH', dominance.type = 'underdominant')

ambiguous.WPH.fractions <- dominanceFractionCalling(heterosis.type = 'WPH', dominance.type = 'ambiguous')
dominant.WPH.fractions <- dominanceFractionCalling(heterosis.type = 'WPH', dominance.type = 'dominant')
homozygous.enrichment.WPH.fractions <- dominanceFractionCalling(heterosis.type = 'WPH', dominance.type = 'homozygous.genotype.enriched')
overdominant.WPH.fractions <- dominanceFractionCalling(heterosis.type = 'WPH', dominance.type = 'overdominant')
underdominant.WPH.fractions <- dominanceFractionCalling(heterosis.type = 'WPH', dominance.type = 'underdominant')

dominance.fractions.all.data <- rbind(dominant.BPH.fractions, overdominant.BPH.fractions, underdominant.BPH.fractions,
                                      homozygous.enrichment.BPH.fractions, dominant.WPH.fractions, overdominant.WPH.fractions,
                                      underdominant.WPH.fractions, homozygous.enrichment.WPH.fractions)

dominance.fractions.all.data <- melt(lapply(split(dominance.fractions.all.data, dominance.fractions.all.data$alpha.threshold),
                                       addDominanceAndHomozygous.genotype.enriched))

##Calculating values for paper
homozygous.genotype.enriched.population.wide <- dominance.call.df[dominance.call.df$dominance == 'homozygous.genotype.enriched' &
                                                     dominance.call.df$heterosis == 'BPH' &
                                                     dominance.call.df$L2 == paste('p', dominance.alpha.value, sep = ''),]
homozygous.genotype.enriched.population.wide <- homozygous.genotype.enriched.population.wide[c(seq(1,length(homozygous.genotype.enriched.population.wide$change), by = 3)), ]

dominance.population.wide <- dominance.call.df[dominance.call.df$dominance == 'dominant' &
                                                     dominance.call.df$heterosis == 'BPH' &
                                                     dominance.call.df$L2 == paste('p', dominance.alpha.value, sep = ''),]
dominance.population.wide <- dominance.population.wide[c(seq(1,length(dominance.population.wide$change), by = 3)), ]

overdominance.population.wide <- dominance.call.df[dominance.call.df$dominance == 'overdominant' &
                                                     dominance.call.df$heterosis == 'BPH' &
                                                     dominance.call.df$L2 == paste('p', dominance.alpha.value, sep = ''),]
overdominance.population.wide <- overdominance.population.wide[c(seq(1,length(overdominance.population.wide$change), by = 3)), ]

number.of.qtls.dominance <- length(dominance.population.wide$change)
number.of.qtls.heterosygous.genotype.enriched <- length(dominance.population.wide$change) + length(homozygous.genotype.enriched.population.wide$change) #this is number of homo, not het
number.of.qtls.overdominance <- length(overdominance.population.wide$change)
```

```{r plot dominance}
plot.dominance.fraction.BPH <- ggplot(data = dominance.fractions.all.data[dominance.fractions.all.data$heterosis.type == 'BPH',],
       aes(alpha.threshold, value, color = dominance.type, group = dominance.type)) +
  scale_color_manual(values = cb.palette[2:8]) +
  geom_line(size = 1.5) + geom_point(size = 2.5) +
  labs(y = 'Frequency', x = 'FDR') +
  theme_bw() +  theme(panel.grid.minor = element_blank()) +
  scale_y_continuous(limits = c(0,0.5))
#svg(filename = '~/Dropbox/myResearch/My_10k/figures/review-NatureCom/dominance-fractions-BPH.svg',
#    width = 7, height = 1.65)

plot.dominance.fraction.WPH <- ggplot(data = dominance.fractions.all.data[dominance.fractions.all.data$heterosis.type == 'WPH',],
       aes(alpha.threshold, value, color = dominance.type, group = dominance.type)) +
  scale_color_manual(values = cb.palette[2:8]) + geom_line(size = 1.5) + geom_point(size = 2.5) +
  labs(y = 'Frequency', x = 'FDR') + theme_bw() +  theme(panel.grid.minor = element_blank()) +
  theme(legend.position = 'none') + scale_y_continuous(limits = c(0,0.6))
svg(filename = '~/Dropbox/myResearch/My_10k/figures/review-NatureCom/dominance-fractions-WPH.svg',
    width = 8, height = 1.5)
```

```{r overdominance in BPH vs entire population}
######
## This chunk looks at whether the QTLs where we found overdominance contributions in the BPH POLs
## also show overdominance when we look at the phenotypic distribution of the entire population
####
overdominance.population.wide <- dominance.call.df[dominance.call.df$dominance == 'overdominant' &
                                                     dominance.call.df$heterosis == 'BPH' &
                                                     dominance.call.df$L2 == paste('p', dominance.alpha.value, sep = ''),]
overdominance.population.wide <- overdominance.population.wide[c(seq(1, length(overdominance.population.wide$change), by = 3)), ]
row.names(overdominance.population.wide) <- paste(overdominance.population.wide$phenotype, overdominance.population.wide$marker, sep = '|')

##Creating genotype phenotype plots for QTLs with significant overdominance in BPH POLs
##I use two of these plots in the paper for figure 4. They are the boxplots in that figure
over.dominance.phenotype.dist <- apply(overdominance.population.wide, MARGIN = 1, 
  FUN = function(x) PhenotypeMarkerDistribution(phenotype = x[['phenotype']],
                                                marker = x[['marker']], output = 'plot'))

##Looking at significant overdominance for entire population at QTLs showing overdominance for BPH POLs
over.dominance.pheno.dist.pvalues <- list()
for(i in names(over.dominance.phenotype.dist)) {
  over.dominance.pheno.dist.pvalues[[i]] <- overDominancePhenoDistT.test(x = over.dominance.phenotype.dist[[i]]$data)
}

##Organizing the list into a data.frame
##In the paper I put the amount of QTLs found with overdominance contributions, and how many of them are with the heterozygotes significantly better
over.dominance.pheno.dist.pvalues.final <- do.call(rbind, over.dominance.pheno.dist.pvalues)
over.dominance.pheno.dist.pvalues.final.result <- over.dominance.pheno.dist.pvalues.final[over.dominance.pheno.dist.pvalues.final$p.value <= dominance.alpha.value
                                                                                           & over.dominance.pheno.dist.pvalues.final$status == 'hetBest',]
```
















